packages/java/endpoint/src/test/java/dev/hilla/EndpointInvokerTest.java

```diff
@@ -0,0 +1,172 @@
+
+    @Autowired
+    private ApplicationContext applicationContext;
+
+    @Mock
+    private EndpointAccessChecker endpointAccessChecker;
+
+    @Mock
+    private EndpointNameChecker endpointNameChecker;
+
+    @Mock
+    private HttpServletRequest requestMock;
+
+    @Mock
+    private Principal principal;
+
+    @Mock
+    private ExplicitNullableTypeChecker explicitNullableTypeChecker;
+
+    @Mock
+    private ServletContext servletContext;
+
+    private EndpointInvoker endpointInvoker;
+    private EndpointRegistry endpointRegistry;
+
+    @Before
+    public void setUp() {
+        when(requestMock.getUserPrincipal()).thenReturn(principal);
+
+        when(endpointNameChecker.check(any())).thenReturn(null);
+
+        endpointRegistry = new EndpointRegistry(endpointNameChecker);
+
+        endpointInvoker = new EndpointInvoker(applicationContext, null,
+                explicitNullableTypeChecker, servletContext, endpointRegistry) {
+            protected EndpointAccessChecker getAccessChecker() {
+                return endpointAccessChecker;
+            }
+        };
+    }
+
+    @Test
+    public void when_invokedMethod_isDeclaredIn_Endpoint_accessCheckingIsDoneBasedOn_EndpointMethod()
+            throws Exception {
+
+        @EndpointExposed
+        class ParentEndpoint {
+            public String sayHello() {
+                return "Hello from ParentEndpoint";
+            }
+        }
+
+        @Endpoint
+        class TestEndpoint extends ParentEndpoint {
+            public void test() {
+            }
+        }
+
+        TestEndpoint test = new TestEndpoint();
+
+        endpointRegistry.registerEndpoint(test);
+
+        endpointInvoker.invoke("TestEndpoint", "test", body, principal,
+                requestMock::isUserInRole);
+
+        Mockito.verify(endpointAccessChecker, Mockito.times(1))
+                .check(any(Method.class), any(), any());
+        Mockito.verify(endpointAccessChecker, Mockito.times(0))
+                .check(any(Class.class), any(), any());
+    }
+
+    @Test
+    public void when_invokedMethod_isDeclaredIn_EndpointExposed_accessCheckingIsDoneBasedOn_EndpointClass()
+            throws Exception {
+
+        @EndpointExposed
+        class ParentEndpoint {
+            public String sayHello() {
+                return "Hello from ParentEndpoint";
+            }
+        }
+
+        @Endpoint
+        class TestEndpoint extends ParentEndpoint {
+            public void test() {
+            }
+        }
+
+        TestEndpoint test = new TestEndpoint();
+
+        endpointRegistry.registerEndpoint(test);
+
+        endpointInvoker.invoke("TestEndpoint", "sayhello", body, principal,
+                requestMock::isUserInRole);
+
+        Mockito.verify(endpointAccessChecker, Mockito.times(0))
+                .check(any(Method.class), any(), any());
+        Mockito.verify(endpointAccessChecker, Mockito.times(1))
+                .check(any(Class.class), any(), any());
```

Comment:
There is some ambiguity regarding which class is used here. Should we assert that it is `TestEndpoint`, and not `ParentEndpoint`?